### PR TITLE
Code monitors: truncate long Slack blocks

### LINKED ISSUE #49979: Code monitors: Slack webhooks fail with `400 Bad Request: invalid blocks`.
[Original thread](https://sourcegraph.slack.com/archives/CHEKCRWKV/p1678366281851429). 

```
non-200 response 400 400 Bad Request with body "invalid_blocks"
```

It sounds like the problem might be that blocks are not truncated and are exceeding the [3000 character limit](https://api.slack.com/reference/block-kit/composition-objects#text) for a text block. We should ensure we always truncate our blocks so that they never exceed 3000 characters.

To get started faster, [this](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@05efe58997a5d8713a33c873130df0f7984025fc/-/blob/enterprise/internal/codemonitors/background/slack.go?L23-23) is where we'll need to make the change.

### PR DESCRIPTION
**Pull request by jtibshirani**
Slack requires that text blocks be no longer than 3000 characters. We already limit the number of lines in the match blocks, but if the lines are long then the block can still exceed the character limit. With this change, we also truncate lines if they will push the block over the character limit.

Fixes #49979.

## Test plan

New unit test for character-based truncation.

Also tested manually -- reproduced the error, then verified this change resolves it.

### CODE REVIEW

**Comment by jtibshirani** [comment]
Here's an example of truncating lines based on the character limit: 
Screenshot 2023-03-28 at 4 50 57 PM

File: enterprise/internal/codemonitors/background/slack.go [hunk 1]
```diff
@@ -85,13 +81,42 @@ func formatCodeBlock(s string) string {
...
 }
 
-// truncateString truncates the input to 10 lines.
-func truncateString(input string) string {
-	const lines = 10
+// truncateMatchContent truncates the result content's 2500 bytes and 10 lines,
+// whichever is more restrictive.
+//
+// We limit the bytes to ensure we don't hit Slack's max block size of 3000
+// characters. To be conservative, we truncate to 2500 bytes. We also limit the
```
Comment (limitedmage) [inline enterprise/internal/codemonitors/background/slack.go:88]: q: Is it possible that multi-byte character (eg. emojis or CJK character) will be split up by using this truncation method? If yes, is there a way we can prevent this?

Comment (jtibshirani) [reply enterprise/internal/codemonitors/background/slack.go:88]: We don't actually truncate the text by byte. We always truncate entire lines -- if a line will push us over the 2500 limit, then we truncate it and all subsequent lines. This seemed simplest and also produces a clean output. I now see this comment is not clear, I can update it :) 

For context, we could use a rune/ character limit instead, but bytes seems safest so we don't count on the Slack API's meaning of "character".

Comment (limitedmage) [reply enterprise/internal/codemonitors/background/slack.go]: Got it, makes sense! So if the first line was already 2500+ characters, no preview will be sent?

Comment (jtibshirani) [reply enterprise/internal/codemonitors/background/slack.go]: We will send a result preview that looks like this:

Screenshot 2023-03-29 at 12 12 59 PM

This is very unlikely to happen in practice though:
* Diff matches begin with the file name, which should be < 2500 ASCII characters
* Commit message matches begin with the (short) title of the commit

Comment (limitedmage) [reply enterprise/internal/codemonitors/background/slack.go]: Awesome, thanks for looking into this, it all sounds very reasonable.

Comment (jtibshirani) [reply enterprise/internal/codemonitors/background/slack.go]: No problem, I had all the same questions when I was developing the fix!

**Review by keegancsmith — APPROVED** [review]

**Pull request merged**